### ISSUE TITLE
Bug with rewriting ImplicationScope wrapped in ExecutionOutputLink

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#2225]
The following

```scheme
(define (dummy x) x)

(Evaluation
   (Predicate "P")
   (Concept "A"))

(define bl
(Bind
   (Present
      (Evaluation
         (Predicate "P")
         (Variable "$X")))
   (ExecutionOutput
      (GroundedSchema "scm: dummy")
      (ImplicationScope
         (Evaluation
            (Predicate "P")
            (Variable "$X"))
         (Evaluation
            (Predicate "Q")
            (Variable "$X"))))))

(cog-execute! bl)
```

wrongly produces

```scheme
      (ImplicationScope
         (Evaluation
            (Predicate "P")
            (Concept "A"))
         (Evaluation
            (Predicate "Q")
            (Variable "$X")))
```

while it should produce

```scheme
      (ImplicationScope
         (Evaluation
            (Predicate "P")
            (Variable "$X"))
         (Evaluation
            (Predicate "Q")
            (Variable "$X")))
```

that is leave untouched the implication scope since `$X` is bound to it.

**Comment by ngeiswei** [comment]
The following branch https://github.com/ngeiswei/atomspace/commits/rewrite-utest contains a unit test for it.

**Comment by linas** [comment]
Yeah, looks like a bug ... in `Instantiator.cc` and/or maybe I guess `Context.cc` ... I guess you are working on this.

**Comment by ngeiswei** [comment]
I found the problem. `FreeVariables::substitute_scoped` does not properly hide the variables in the scope in that instance. The thing is that in that `FreeVariables::substitute_scoped` is called with variables that aren't as `beta_reduce` in `Instantiator.cc` construct a `FreeVariables` including variables and non variables in vmap, see

https://github.com/opencog/atomspace/blob/fea73cf3582d0aa993032f1e0dcd06a893623dbf/opencog/atoms/execution/Instantiator.cc#L55-L56

and so `FreeVariables::substitute_scoped` fails to recognize that

```scheme
(Evaluation
  (Predicate "P")
  (Variable "$X"))
```

should not be substituted by

```scheme
(Evaluation
  (Predicate "P")
  (Concept "A"))
```

I see 2 ways to fix that
1. Fix `beta_reduce` so that `crub` only considers variables.
2. Fix `FreeVariables::substitute_scoped` so that it robustly supports non-variables.

I think I will go with 2. as to make `FreeVariables::substitute_scoped` more robust in general, and I suppose we want to take advantage of doing whole subtree substitutions rather than just variables, which is potentially faster, thus restricting `crub` to be just variables, even though I suppose it might create future troubles.
